src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala

```diff
@@ -0,0 +1,294 @@
+      if (receiptsQueue.nonEmpty) requestReceipts(peer)
+      else if (blockBodiesQueue.nonEmpty) requestBlockBodies(peer)
+      else if (context.child(blockHeadersHandlerName).isEmpty) requestBlockHeaders(peer)
+      else if (nodesQueue.nonEmpty) requestNodes(peer)
+    }
+
+    def requestReceipts(peer: ActorRef): Unit = {
+      val receipts = receiptsQueue.take(receiptsPerRequest)
+      val handler = context.actorOf(FastSyncReceiptsRequestHandler.props(peer, receipts, receiptStorage))
+      context watch handler
+      assignedHandlers += (handler -> peer)
+      receiptsQueue = receiptsQueue.drop(receiptsPerRequest)
+    }
+
+    def requestBlockBodies(peer: ActorRef): Unit = {
+      val blockBodies = blockBodiesQueue.take(blockBodiesPerRequest)
+      val handler = context.actorOf(FastSyncBlockBodiesRequestHandler.props(peer, blockBodies, blockBodiesStorage))
+      context watch handler
+      assignedHandlers += (handler -> peer)
+      blockBodiesQueue = blockBodiesQueue.drop(blockBodiesPerRequest)
+    }
+
+    def requestBlockHeaders(peer: ActorRef): Unit = {
+      val handler = context.actorOf(FastSyncBlockHeadersRequestHandler.props(
+        peer,
+        nodeStatusHolder().blockchainStatus.bestNumber + 1,
+        blockHeadersPerRequest,
+        nodeStatusHolder, blockHeadersStorage), blockHeadersHandlerName)
+      context watch handler
+      assignedHandlers += (handler -> peer)
+    }
+
+    def requestNodes(peer: ActorRef): Unit = {
+      val nodes = nodesQueue.take(nodesPerRequest)
+      val handler = context.actorOf(FastSyncNodesRequestHandler.props(peer, nodes, evmStorage, mptNodeStorage))
+      context watch handler
+      assignedHandlers += (handler -> peer)
+      nodesQueue = nodesQueue.drop(nodesPerRequest)
+    }
+
+    def unassignedPeers: Set[ActorRef] = peersToDownloadFrom diff assignedHandlers.values.toSet
+
+    def anythingQueued: Boolean =
+      nodesQueue.nonEmpty ||
+      blockBodiesQueue.nonEmpty ||
+      receiptsQueue.nonEmpty
+
+    def fullySynced: Boolean =
+      nodeStatusHolder().blockchainStatus.bestNumber == targetBlock.number &&
+      !anythingQueued &&
+      assignedHandlers.isEmpty
+  }
+}
+
+object FastSyncController {
+  def props(peerManager: ActorRef, nodeStatusHolder: Agent[NodeStatus], mptNodeStorage: MptNodeStorage,
+            blockHeadersStorage: BlockHeadersStorage, blockBodiesStorage: BlockBodiesStorage,
+            receiptStorage: ReceiptStorage, evmStorage: EvmCodeStorage): Props =
+    Props(new FastSyncController(peerManager, nodeStatusHolder, mptNodeStorage, blockHeadersStorage,
+      blockBodiesStorage, receiptStorage, evmStorage))
+
+  case class StartFastSync(targetBlockHash: ByteString)
+
+  case class EnqueueNodes(hashes: Seq[HashType])
+  case class EnqueueBlockBodies(hashes: Seq[ByteString])
+  case class EnqueueReceipts(hashes: Seq[ByteString])
+
+  case class UpdateDownloadedNodesCount(update: Int)
+
+  trait HashType {
```

Comment:
Maybe this trait should be `sealed`? After upgrading to Scala 2.12 the compiler warns me of match not exhaustive in `handleMptDownload` if this trait isn't sealed.